server/aesdsocket.c

```diff
@@ -0,0 +1,301 @@
+
+    if (signal(SIGINT, sigint_handler) == SIG_ERR)
+    {
+        perror("signal");
+        printf("Couldn't handle SIGINT");
+        cleanup();
+        exit(-1);
+    }
+    if (signal(SIGTERM, sigint_handler) == SIG_ERR)
+    {
+        perror("signal");
+        printf("Couldn't handle SIGTERM");
+        cleanup();
+        exit(-1);
+    }
+
+    socket_fd = socket(AF_INET, SOCK_STREAM, 0);
+    if (socket_fd == -1)
+    {
+        perror("socket");
+        printf("Socket couldn't be created\n");
+        cleanup();
+        exit(-1);
+    }
+
+    if (setsockopt(socket_fd, SOL_SOCKET, SO_REUSEADDR, &reuseaddr, sizeof(int)) == -1)
+    {
+        perror("setsockopt");
+        exit(-1);
+    }
+
+    memset(&server, 0, sizeof(server));
+
+    server.sin_family = AF_INET;
+    server.sin_addr.s_addr = htonl(INADDR_ANY);
+    server.sin_port = htons(PORT);
+
+    if ((bind(socket_fd, (struct sockaddr *)&server, sizeof(server))) < 0)
+    {
+        perror("bind");
+        printf("Socket bind failed\n");
+        cleanup();
+        exit(-1);
+    }
+
+    if (useDaemon)
+    {
+        pid = fork();
+        if (pid == -1)
+        {
+            perror("fork");
+            printf("Fork failed\n");
+            cleanup();
+            exit(-1);
+        }
+        else if (pid != 0)
+        {
+            cleanup();
+            exit(EXIT_SUCCESS);
+        }
+    }
+    if ((listen(socket_fd, 10)) != 0)
+    {
+        perror("listen");
+        printf("Listen failed\n");
+        cleanup();
+        exit(-1);
+    }
+
+    client_size = sizeof(client);
+
+    fd = open("/var/tmp/aesdsocketdata", O_RDWR | O_CREAT | O_APPEND, 0644);
+    while (1)
+    {
+        conn_fd = accept(socket_fd, (struct sockaddr *)&client, &client_size);
+        if (conn_fd == -1)
+        {
+            perror("accept");
+            printf("Accept failed\n");
+            cleanup();
+            exit(-1);
+        }
+
+        isConnectionInProgress = true;
+
+        inet_ntop(AF_INET, &(client.sin_addr), client_str, INET_ADDRSTRLEN);
+
+        syslog(LOG_DEBUG, "Accepted connection from %s", client_str);
+
+        buflen = START_LEN;
+        pos = 0;
+        buffer = (char *)malloc(sizeof(char) * buflen);
+        if (buffer == NULL)
+        {
+            perror("malloc");
+            printf("buffer malloc failed\n");
+            cleanup();
+            exit(-1);
+        }
+        memset(buffer, 0, buflen);
+        do
+        {
+            len_recv = recv(conn_fd, buffer + pos, buflen - pos - 1, 0);
+            if (len_recv == -1)
+            {
+                perror("recv");
+                printf("Receive failed\n");
+                cleanup();
+                exit(-1);
+            }
+            else if (len_recv == buflen - pos - 1)
+            {
+                buflen += START_LEN;
+                pos = buflen - START_LEN - 1;
+                temp = (char *)realloc(buffer, buflen);
```

Comment:
Think if the string/ data is way more than heap size? How would you handle in that case?